agent/src/lib/helpers.ts

```diff
@@ -9,10 +17,121 @@ export function wildcardMatch(target: string, pattern: string): boolean {
     return exp.test(target);
 }
 
+/**
+ * Enhanced debug printing. For XPC dicitionaries we print entire byte arrays instead of just the beginning. This enhances the output compared to the standard output by debugPrinting 
+ * @param ptr Objective C Pointer 
+ * @returns A debug string with all raw data contents 
+ */
 export function objcObjectDebugDesc(ptr: NativePointer) {
-    return (new ObjC.Object(ptr)).toString();
+    // We enhance the debug description by actually printing complete data blobs as base64 instead of trimming the data when using the current implementation 
+
+    let objcObject = new ObjC.Object(ptr);
+    let className = objcObject.$className;  
+    let outString = ""
+
+
+    if (className === "OS_xpc_dictionary") {
+        let entries = xpcDictionaryGetCount.call(objcObject); 
+
+        if (entries > 0) {
+            outString += debugDescriptionForXPCDictionary(objcObject, entries); 
+        }else {
+            outString += objcObject.toString(); 
+        }
+    }else {
+        outString += objcObject.toString(); 
+    }
+
+    return outString;
+}
+
+export function debugDescriptionForXPCDictionary(xpcDict: ObjC.Object, count) {
+
+    let outString = "<OS_xpc_dictionary> { count = " + count + " "; 
+    outString += "contents = \n\t"
+
+    /**
+     * See: https://developer.apple.com/documentation/xpc/1505404-xpc_dictionary_apply?language=objc
+     */
+     const block_impl = function(key: NativePointer, value: NativePointer): boolean {
+        const valueType = objcObjectDebugDesc(<NativePointer>xpcGetType.call(value));
+        
+        let keyString = key.readCString(); 
+        outString += "\"" + keyString + "\" => "; 
+
+        let objcValue = new ObjC.Object(value); 
+
+        switch (valueType) {
+            case 'OS_xpc_dictionary':
+                let entriesCount = xpcDictionaryGetCount.call(value); 
+                outString += debugDescriptionForXPCDictionary(objcValue, entriesCount); 
+                break;
+            case 'OS_xpc_data':
+                const bytesPtr = <NativePointer>xpcDataGetBytesPtr.call(value);
+                const length = <NativePointer>xpcDataGetLength.call(value); 
+                let hexString = hexStringForBytes(bytesPtr,length); 
+                // let hexString = "empty"; 
+                outString += `<data> { length = ${length.valueOf()} bytes, contents = \n\t\t${hexString}\n\t\t}\n\t`; 
+                break;
+            default:
+                outString += objcValue.toString() + "\n\t"; 
+                break;
+        }
+        return true;
+    }
+
+    const applierBlock = new ObjC.Block({
+        implementation: block_impl,
+        retType: 'bool',
+        argTypes: ['pointer', 'pointer']
+    });
+
+    xpcDictionaryApply.call(xpcDict, applierBlock.handle);
+
+    outString += "\n}"
+
+    return outString; 
 }
 
+
+export function objcDebugDescriptionNSData(data: ObjC.Object) {
```

Comment:
It seems that this function is never used.

Reply:
Yes. I initially thought the XPC dictionary would contain NSData instead of OS_xpc_data. This is the function I used to parse NSData.